drop-transfer/src/ws/client/socket.rs

```diff
@@ -0,0 +1,56 @@
+use std::{
+    io,
+    time::{Duration, Instant},
+};
+
+use futures_util::{SinkExt, StreamExt};
+use tokio::net::TcpStream;
+use tokio_tungstenite::{tungstenite::Message, WebSocketStream};
+
+pub type WsStream = WebSocketStream<TcpStream>;
+
+pub struct WebSocket {
+    stream: WsStream,
+    recv_last: Instant,
+    recv_timeout: Duration,
+    send_timeout: Duration,
+}
+
+impl WebSocket {
+    pub fn new(stream: WsStream, recv_timeout: Duration, send_timeout: Duration) -> Self {
+        Self {
+            stream,
+            recv_last: Instant::now(),
+            recv_timeout,
+            send_timeout,
+        }
+    }
+
+    pub async fn send(&mut self, msg: Message) -> crate::Result<()> {
+        tokio::time::timeout(self.send_timeout, self.stream.send(msg))
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::TimedOut, err))??;
+
+        Ok(())
+    }
+
+    pub async fn recv(&mut self) -> crate::Result<Message> {
+        let timeout = self.recv_timeout.saturating_sub(self.recv_last.elapsed());
+
+        let msg = tokio::time::timeout(timeout, self.stream.next())
+            .await
+            .map_err(|err| io::Error::new(io::ErrorKind::TimedOut, err))?
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Emptry socket stream"))??;
+
+        Ok(msg)
```

Comment:
Similar question to above, if the `await` resulted in a timeout error we map it to `TimedOut` but why do we need another `ok_or_else` for?